File: back-end/src/main/java/com/mgimss/mgimss/classes/Appliance.java
package com.mgimss.mgimss.classes;


import javax.persistence.*;
import java.io.Serializable;


@Entity
@Table(name= "appliance")
@IdClass(Appliance.ApplianceId.class)
public class Appliance {



    @Id
    @Column(name = "id")
    private int id;
    @Id
    @Column(name = "time")
    private String time;
    @Id
    @Column(name="username")
    private String username;

    private String name;
    private float voltage;
    private float current;
    private int status;

    public static class ApplianceId implements Serializable{
        private int id;
        private String time;
        private String username;

        public String getUsername() {
            return username;
        }

        public void setUsername(String username) {
            this.username = username;
        }

        public String getTime() {
            return time;
        }

        public int getId() {
            return id;
        }

        public void setTime(String time) {
            this.time = time;
        }

        public void setId(int id) {
            this.id = id;
        }
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getTime() {
        return time;
    }

    public int getId() {
        return id;
    }

    public void setTime(String time) {
        this.time = time;
    }

    public void setId(int id) {
        this.id = id;
    }
    public float getCurrent() {
        return current;
    }

    public void setCurrent(float current) {
        this.current = current;
    }

    public float getVoltage() {
        return voltage;
    }

    public void setVoltage(float voltage) {
        this.voltage = voltage;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: Appliance_Simulator/appliance.py
import serial
ser = serial.Serial('COM3', 9600)
option_on = b'b'
option_off = b'a'


class Appliance:
    def __init__(self, app_id, name="unknown appliance", voltage=220, current=0, status=0):
        self.__id = app_id
        self.__voltage = voltage
        self.__current = current
        self.__name = name
        # status==1 for on and 0 for off.
        self.__status = status

    def get_id(self):
        return self.__id

    def set_voltage(self, voltage):
        self.__voltage = voltage

    def set_current(self, current):
        self.__current = current

    def set_name(self, name):
        self.__name = name

    def turn_on(self):
        if self.__status == 1:
            return -1
        self.__status = 1
        return 0

    def turn_off(self):
        if self.__status == 0:
            return -1
        self.__status = 0
        return 0

    def get_voltage(self):
        return self.__voltage

    def get_current(self):
        return self.__current

    def get_name(self):
        return self.__name

    def get_status(self):
        # 1 for on and 0 for off
        return self.__status

    def get_all(self):
        # get the name, voltage, current and status of the appliance
        return self.get_id(), self.get_name(), self.get_voltage(),\
               self.get_current(), self.get_status()


class ArduinoLED(Appliance):
    def __init__(self, app_id, name="unknown appliance", voltage=220, current=0, status=0):
        Appliance.__init__(self, app_id, name, voltage, current, status)
        if status == 1:
            ser.write(option_on)
        if status == 0:
            ser.write(option_off)

    def turn_on(self):
        ser.write(option_on)
        return Appliance.turn_on(self)

    def turn_off(self):
        ser.write(option_off)
        return Appliance.turn_off(self)
